# Hw 8


## Question 1.1

1a. Securities Traded:
Types: Fixed-income securities (e.g., U.S. Treasury bonds, Japanese Government Bonds, corporate bonds), derivatives (interest rate swaps, options), mortgage-backed securities (MBS), and equity-related trades (e.g., Royal Dutch/Shell arbitrage).
Focus: Instruments with mispriced spreads or structural inefficiencies, often involving low credit risk (e.g., government-backed securities).

b. Trading Frequency:
Long-term Horizon: Most trades had holding periods ranging from 6 months to 2 years or longer.
Low Turnover: LTCM's positions often required significant time for spreads or valuations to converge.

c. Skewness:
Many Small Wins: LTCM’s strategies focused on capturing small mispricings with high probability of success (e.g., convergence trades with predictable outcomes). However, the use of leverage amplified these small gains into substantial profits.

d. Forecasting:
Quantitative Models: LTCM relied on mathematical models to identify mispricings in financial markets and predict the reversion of spreads to normal levels.
Institutional Behavior: Many trades were based on understanding institutional demand (e.g., regulatory constraints or liquidity needs) that created temporary pricing inefficiencies.

## Question 1.2

Long-Term Capital Management (LTCM) held significant advantages over competitors due to its expertise, quantitative methodologies, and strategic relationships. Led by John Meriwether and a team of renowned quantitative experts and Nobel laureates, LTCM leveraged unparalleled intellectual capital to develop proprietary models that allowed it to exploit subtle mispricings and systematically manage risk. These models, grounded in advanced mathematical and statistical theories, formed the backbone of LTCM’s strategy and gave the firm a significant edge in financial markets.

Proprietary Models and Applications
Prepayment Risk in Mortgage-Backed Securities (MBS):

LTCM developed a sophisticated model to analyze prepayment risks, which arise when homeowners refinance their mortgages during periods of falling interest rates.
Their model estimated the fair value of MBS securities and their sensitivity to interest rate changes, enabling LTCM to capitalize on mispricings overlooked by less sophisticated investors. For example, their analysis allowed them to predict how interest rate volatility would influence refinancing behavior, adjusting their positions accordingly.
Stock Price Dynamics:

LTCM employed stochastic models to predict stock price movements and measure exposure to market risks. By using methods like the Black-Scholes option pricing framework, they dynamically adjusted their positions in derivatives to exploit volatility mispricings while minimizing directional exposure to the underlying assets.
Interest Rate Models:

For fixed-income arbitrage, LTCM used term structure models, such as the Heath-Jarrow-Morton framework, to predict changes in the yield curve and identify relative value trades. These models allowed LTCM to take advantage of anomalies, such as the "butterfly spread" trades, where mid-term rates deviated from short-term and long-term rates.
Swap Spread Trades:

LTCM employed models to assess the fair value of swap spreads. For example, in Japanese Government Bonds (JGBs), their models evaluated spreads between bond yields and swap fixed rates while incorporating funding costs and market liquidity conditions. This enabled LTCM to predict mean-reversion and profit from misaligned spreads.
Volatility Models:

In its volatility-selling strategy, LTCM used models to compare implied volatility (used to price options) with realized volatility in the markets. For instance, when implied volatility of equity index options exceeded historical volatility, their models indicated a high probability of profits by selling options and dynamically hedging them.
Institutional Demand-Driven Mispricings:

LTCM modeled pricing anomalies created by institutional constraints such as regulatory requirements or operational inefficiencies. For example:
They exploited the pricing differential between Royal Dutch Petroleum and Shell, which traded at different valuations despite being identical in economic terms.
LTCM profited from market imbalances caused by banks and governments hedging large positions, such as in the European bond markets during the transition to the Euro.
Strategic Strengths
Low Financing Costs:

LTCM negotiated favorable financing terms, including low repo rates and long-term funding agreements. This efficiency allowed them to scale trades with minimal cost, amplifying returns on positions with small mispricings. For example, LTCM financed U.S. Treasury positions at LIBOR - 20 basis points, reducing carry costs and increasing net profitability.
Scalable Risk-Taking:

LTCM’s ability to use leverage at a ratio of 19:1 amplified the impact of small pricing discrepancies. For instance, a 1-basis-point change in swap spreads on a $1 billion notional trade generated $100,000 in profit, and LTCM’s portfolio was structured to maximize these incremental gains at scale.
Uncorrelated Strategies:

LTCM’s diversified portfolio spanned multiple asset classes and geographic regions. By incorporating uncorrelated risks, such as MBS prepayment risk, European bond convergence trades, and equity arbitrage, LTCM reduced overall portfolio volatility while maintaining high returns.
Local Market Insights:

LTCM partnered with institutional investors globally, gaining access to local perspectives on macroeconomic policies and market behaviors. These partnerships informed their strategies in regional markets, such as Japanese swap spreads or European government bonds, ensuring well-informed trades with minimized risk.
Quantitative Portfolio Construction:

LTCM applied advanced portfolio optimization techniques, balancing risk and return while ensuring zero beta exposure to broad market risks (e.g., equities, bonds, and currencies). This approach focused on maximizing returns from idiosyncratic mispricings rather than relying on directional market movements.

## Question 1.3

a. Collateral Haircuts:
Risk: Financing counterparties may demand additional collateral (e.g., higher haircuts) during market stress.
Management:
LTCM estimated worst-case haircuts and structured long-term repo agreements to minimize sudden demands for collateral.

b. Repo Maturity:
Risk: Short-term repos need frequent rolling over, exposing LTCM to market disruptions.
Management:
LTCM relied on term repos with maturities of 6–12 months instead of overnight repos to avoid frequent rollovers.

c. Equity Redemption:
Risk: Investors withdrawing capital could force LTCM to liquidate positions prematurely.
Management:
LTCM staggered redemption rights (e.g., only one-third of equity could be redeemed annually) and had a three-year lock-up period for new investors.

d. Loan Access:
Risk: Loss of access to credit lines during crises.
Management:
LTCM negotiated a $700 million unsecured credit line with no discretionary clauses, ensuring availability unless NAV dropped significantly.


## Question 1.4
Short-Term Liquidity Risk:

LTCM measured the short-term volatility of prices due to traders’ liquidity needs, assuming these were temporary and unrelated to fundamental value.
For example, the monthly standard deviation of U.S. Treasury swap spreads was estimated at 4 basis points, higher than the long-term annualized estimate of 7 basis points.
Economic Stress Testing:

LTCM modeled liquidity needs during extreme market events (e.g., widening repo haircuts, forced liquidations).
Correlation Assumptions:

For short-term risk, LTCM conservatively assumed that its trades were positively correlated due to other market participants unwinding similar positions.

## Question 1.5

Current Leverage:
As of 1997, LTCM had a leverage ratio of 19:1 (total assets to equity capital). This was lower than its historical range (19:1 to 31:1) and comparable to large banks.

Risk Assessment:

LTCM's use of long-short positions ensured exposure to specific risks (e.g., spread movements) while mitigating broad market risks (e.g., interest rate changes). LTCM tried to mainatin beta neutral in stcoks, bonds, and other asset classes.

## Question 1.6


a. Diverging spreads could lead to mark-to-market losses, requiring additional collateral and possibly forcing LTCM to exit trades at a loss. Liquidity risk occurs when you cannot maintain or exit positions due to:Insufficient cash or collateral to meet obligations (e.g., margin calls). An inability to sell assets without significantly impacting their market price. Temporary widening spreads can force LTCM to sell assets to meet margin calls, even when they expect spreads to eventually converge. If LTCM’s positions experience mark-to-market losses, counterparties may demand additional collateral (margin calls) to maintain funding.

b. Leverage Amplification:
High leverage magnifies the impact of temporary spread widening, risking catastrophic losses during extreme market disruptions.

c. Funding Risk:
During market stress, financing costs could rise (e.g., repo rates increase or haircuts widen), making it harder to hold positions. LTCM used repos to finance large positions, including their JGB swap spread trades. If repo rates rose or haircuts widened, LTCM would face higher funding costs or liquidity drains.

d. Unexpected Events:
Structural market shifts or unforeseen regulatory changes could prevent spreads from converging (e.g., a permanent breakdown in market efficiency). LTCM’s European bond convergence trades assumed that spreads between sovereign bonds of different countries (e.g., Italy vs. Germany) would narrow as the countries moved toward adopting the Euro. However, political or economic instability in a member country could have disrupted this convergence, leading to permanent losses.

## Question 2.1: Summary stats


In [17]:
import numpy as np
import os
from arch import arch_model
import re
import sys
import matplotlib.pyplot as plt
from scipy.stats import norm
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats
import os
import sys
from typing import Tuple, Dict, Union, Callable
notebook_dir = "practice"
if os.path.basename(os.getcwd()) == notebook_dir:
    os.chdir(os.pardir)  # Change to parent directory
    sys.path.append(os.getcwd())  # Add the new current directory to sys.path

import cmds.portfolio_management_helper as pmh
import functions.functions_andy as fa
print(os.getcwd())

from sklearn.linear_model import LinearRegression
from dask.distributed import Client, wait
# From HW1
import statsmodels.api as sm
from statsmodels.regression.rolling import RollingOLS
from sklearn.metrics import r2_score
from scipy.stats import describe



c:\Users\Andik\OneDrive\Desktop\Chicago\Portfolio theory and risk management\githubrepo\finm-portfolio-2024


In [52]:
import pandas as pd

# Read data from Excel for LTCM
ltcm = pd.read_excel(
    "./data/ltcm_exhibits_data.xlsx",
    sheet_name="Exhibit 2",
    skiprows=2,
    index_col=0,
    parse_dates=[0]
).iloc[:-4, :]
ltcm.index.name = "Date"

# Convert index to period end and rename columns
ltcm.index = pd.to_datetime(ltcm.index).to_period('M').to_timestamp('M')
ltcm = ltcm[["Net Monthly Performanceb", "Gross Monthly Performancea"]].dropna()
ltcm = ltcm.rename(
    columns={
        "Net Monthly Performanceb": "LTCM Net",
        "Gross Monthly Performancea": "LTCM Gross",
    }
)

# Read SPY returns
spy = pd.read_excel(
    "./data/gmo_analysis_data.xlsx",
    sheet_name="total returns",
    index_col=0,
    parse_dates=[0]
)[["SPY"]]

# Read risk-free rate and adjust for daily rate to monthly
us3m = (
    pd.read_excel(
        "./data/gmo_analysis_data.xlsx",
        sheet_name="risk-free rate",
        index_col=0,
        parse_dates=[0]
    )[["TBill 3M"]] / 252
)

# Convert indices of SPY and US3M
spy.index = spy.index.to_period('M').to_timestamp('M')
us3m.index = us3m.index.to_period('M').to_timestamp('M')

# Join data on month and year of the index
df = ltcm.join(spy, how="inner").join(us3m, how="inner")

# Convert to EXCESS returns
df.loc[:, ["LTCM Net", "LTCM Gross", "SPY"]] = df.loc[:, ["LTCM Net", "LTCM Gross", "SPY"]].subtract(df["TBill 3M"], axis=0)

# Calculate performance metrics (assuming 'pmh' is a module or defined previously)
pmh.calc_summary_statistics(df[["LTCM Net", "LTCM Gross", "SPY"]],annual_factor=12,provided_excess_returns=True).T


LTCM Net           LTCM Gross  \
Mean                                             0.0171               0.0243   
Annualized Mean                                  0.2048               0.2915   
Vol                                              0.0323               0.0394   
Annualized Vol                                   0.1119               0.1363   
Sharpe                                           0.5282               0.6171   
Annualized Sharpe                                1.8299               2.1377   
Min                                             -0.1012              -0.1012   
Max                                              0.0838               0.1158   
Skewness                                        -0.7867              -0.2849   
Excess Kurtosis                                  2.4237               1.2331   
Historical VaR (5.00%)                          -0.0226              -0.0266   
Annualized Historical VaR (5.00%)               -0.0783              -0.0921   
Historical CVaR (5.00%)                         -0.0649              -0.0692   
Annualized Historical CVaR (5.00%)              -0.2247              -0.2397   
Max Drawdown                                    -0.1591              -0.1616   
Peak                                1998-04-30 00:00:00  1998-04-30 00:00:00   
Bottom                              1998-07-31 00:00:00  1998-06-30 00:00:00   
Tangency Weights                                35.1473             -31.4495   
Recovery                                            NaT                  NaT   
Duration (days)                                     NaN                  NaN   
LTCM Net Correlation                             1.0000               0.9929   
LTCM Gross Correlation                           0.9929               1.0000   
SPY Correlation                                  0.1402               0.1502   

                                                    SPY  
Mean                                             0.0186  
Annualized Mean                                  0.2234  
Vol                                              0.0326  
Annualized Vol                                   0.1129  
Sharpe                                           0.5714  
Annualized Sharpe                                1.9792  
Min                                             -0.0520  
Max                                              0.0791  
Skewness                                        -0.4205  
Excess Kurtosis                                 -0.4871  
Historical VaR (5.00%)                          -0.0433  
Annualized Historical VaR (5.00%)               -0.1499  
Historical CVaR (5.00%)                         -0.0477  
Annualized Historical CVaR (5.00%)              -0.1652  
Max Drawdown                                    -0.0520  
Peak                                1997-07-31 00:00:00  
Bottom                              1997-08-31 00:00:00  
Tangency Weights                                -2.6978  
Recovery                            1997-11-30 00:00:00  
Duration (days)                                 91.0000  
LTCM Net Correlation                             0.1402  
LTCM Gross Correlation                           0.1502  
SPY Correlation                                  1.0000

In [59]:
data_2 = pmh.read_excel_default(excel_name="./data/ltcm_exhibits_data.xlsx",sheet_name="Exhibit 2",skiprows=2).dropna()
data_4 = pmh.read_excel_default(excel_name="./data/ltcm_exhibits_data.xlsx",sheet_name="Exhibit 4",skiprows=2).dropna()
data_2.index = pd.to_datetime(data_2.index)
data_4.index = pd.to_datetime(data_4.index)
display(data_2.head())
display(data_4.head())
gmo_spy = pmh.read_excel_default(excel_name="./data/gmo_analysis_data.xlsx",sheet_name="total returns").loc[:,["SPY"]].dropna()
display(gmo_spy.head())
gmo_rf = pmh.read_excel_default(excel_name="./data/gmo_analysis_data.xlsx",sheet_name="risk-free rate").dropna()/12
display(gmo_rf.head())

Fund Capital ($billions)  Gross Monthly Performancea  \
1994-03-01                    1.1000                     -0.0110   
1994-04-01                    1.1000                      0.0140   
1994-05-01                    1.2000                      0.0680   
1994-06-01                    1.2000                     -0.0390   
1994-07-01                    1.4000                      0.1160   

            Net Monthly Performanceb  Index of Net Performance  
1994-03-01                   -0.0130                    0.9900  
1994-04-01                    0.0080                    1.0000  
1994-05-01                    0.0530                    1.0500  
1994-06-01                   -0.0290                    1.0200  
1994-07-01                    0.0840                    1.1000

Total Assets ($billions)  Total Capital ($billions)  \
Month                                                             
1994-06-01                   17.4000                     1.2000   
1994-07-01                   18.6000                     1.4000   
1994-08-01                   23.0000                     1.5000   
1994-09-01                   25.8000                     1.5000   
1994-10-01                   26.2000                     1.5000   

            Leverage Ratio  Capital Ratio  
Month                                      
1994-06-01         14.5000         0.0690  
1994-07-01         13.7000         0.0730  
1994-08-01         15.7000         0.0640  
1994-09-01         17.7000         0.0560  
1994-10-01         17.7000         0.0570

SPY
date              
1993-02-28  0.0107
1993-03-31  0.0224
1993-04-30 -0.0256
1993-05-31  0.0270
1993-06-30  0.0037

TBill 3M
date                
1989-12-31    0.0064
1990-01-31    0.0065
1990-02-28    0.0066
1990-03-31    0.0065
1990-04-30    0.0065

In [60]:
data_2.index = data_2.index.to_period('M')
gmo_rf.index = gmo_rf.index.to_period('M')
gmo_spy.index = gmo_spy.index.to_period('M')
full_data = data_2.merge(gmo_rf, left_index=True, right_index=True, how="inner")
full_data = full_data.merge(gmo_spy, left_index=True, right_index=True, how="inner")
full_data.head()

Fund Capital ($billions)  Gross Monthly Performancea  \
1994-03                    1.1000                     -0.0110   
1994-04                    1.1000                      0.0140   
1994-05                    1.2000                      0.0680   
1994-06                    1.2000                     -0.0390   
1994-07                    1.4000                      0.1160   

         Net Monthly Performanceb  Index of Net Performance  TBill 3M     SPY  
1994-03                   -0.0130                    0.9900    0.0031 -0.0419  
1994-04                    0.0080                    1.0000    0.0034  0.0112  
1994-05                    0.0530                    1.0500    0.0034  0.0159  
1994-06                   -0.0290                    1.0200    0.0036 -0.0229  
1994-07                    0.0840                    1.1000    0.0037  0.0323

In [61]:
full_data.dropna(inplace=True)
sum_stats_data = pd.DataFrame(index = full_data.index)
sum_stats_data.assign()
sum_stats_data = sum_stats_data.assign(gross_excess = full_data["Gross Monthly Performancea"] - full_data["TBill 3M"],
                                       net_excess = full_data["Net Monthly Performanceb"] - full_data["TBill 3M"],
                                       spy_excess = full_data["SPY"] - full_data["TBill 3M"])
display(sum_stats_data.head())
summary_statistics = pmh.calc_summary_statistics(returns=sum_stats_data,
                                                 annual_factor=12,
                                                 provided_excess_returns=True,
                                                 var_quantile=0.05)
display(summary_statistics.T)

gross_excess  net_excess  spy_excess
1994-03       -0.0141     -0.0161     -0.0450
1994-04        0.0106      0.0046      0.0078
1994-05        0.0646      0.0495      0.0125
1994-06       -0.0426     -0.0326     -0.0265
1994-07        0.1123      0.0803      0.0286

Could not convert "date" index to datetime.date
Cannot calculate "Drawdown Duration" calculation because there was no recovery or because index are not dates: 'pandas._libs.tslibs.offsets.MonthEnd' object has no attribute 'days'


gross_excess net_excess spy_excess
Mean                                     0.0203     0.0131     0.0146
Annualized Mean                          0.2433     0.1566     0.1752
Vol                                      0.0393     0.0323     0.0325
Annualized Vol                           0.1363     0.1118     0.1127
Sharpe                                   0.5155     0.4044     0.4489
Annualized Sharpe                        1.7858     1.4009     1.5551
Min                                     -0.1051    -0.1051    -0.0559
Max                                      0.1123     0.0803     0.0750
Skewness                                -0.2788    -0.7811    -0.4162
Excess Kurtosis                          1.2421     2.4336    -0.4852
Historical VaR (5.00%)                  -0.0304    -0.0264    -0.0466
Annualized Historical VaR (5.00%)       -0.1052    -0.0914    -0.1616
Historical CVaR (5.00%)                 -0.0730    -0.0686    -0.0517
Annualized Historical CVaR (5.00%)      -0.2528    -0.2377    -0.1791
Max Drawdown                            -0.1688    -0.1754    -0.0559
Peak                                    1998-04    1997-12    1997-07
Bottom                                  1998-06    1998-07    1997-08
Tangency Weights                        -9.5175    10.9776    -0.4601
Recovery                                    NaT        NaT    1997-12
gross_excess Correlation                 1.0000     0.9929     0.1486
net_excess Correlation                   0.9929     1.0000     0.1382
spy_excess Correlation                   0.1486     0.1382     1.0000

In [ ]:
data = sum_stats_data.loc[:,["net_excess","spy_excess"]]
model = LinearRegression(fit_intercept=True)
model.fit(X = data["spy_excess"].values.reshape(-1,1),y = data["net_excess"].values)
print(model.coef_[0],model.intercept_)
print(r2_score(y_true = data["net_excess"],y_pred = model.predict(data["spy_excess"].values.reshape(-1,1))))

print("Non_linear regression")

from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.pipeline import Pipeline

model = Pipeline([('poly', PolynomialFeatures(degree=2 ,include_bias=False)),('linear', LinearRegression(fit_intercept=True))])
model.fit(X = data["spy_excess"].values.reshape(-1,1),y = data["net_excess"].values)
print(model.named_steps["linear"].coef_,model.named_steps["linear"].intercept_)
print(r2_score(y_true = data["net_excess"],y_pred = model.predict(data["spy_excess"].values.reshape(-1,1))))




0.13710069957665924 0.011050285556661677
0.01909664793554089
Non_linear regression
[ 0.16732139 -1.92887367] 0.01302328153933649
0.0245391473618779


In [68]:
pmh.calc_iterative_regression(multiple_y=data.loc[:,["net_excess"]],X=data.loc[:,["spy_excess"]],annual_factor=12)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Alpha  Annualized Alpha  R-Squared  spy_excess Beta  \
net_excess 0.0111            0.1326     0.0191           0.1371   

            Treynor Ratio  Annualized Treynor Ratio  Information Ratio  \
net_excess         0.0952                    1.1424             0.3457   

            Annualized Information Ratio  Tracking Error  \
net_excess                        1.1975          0.0320   

            Annualized Tracking Error  Fitted Mean  Annualized Fitted Mean  
net_excess                     0.1107       0.0131                  0.1566

In [69]:
k1 = 0.03
k2 = - 0.03
data = data.assign(call_factor = (data["spy_excess"] - k1).apply(lambda x: max(x,0)),
                   put_factor = (k2 - data["spy_excess"]).apply(lambda x: max(x,0)))
model = LinearRegression(fit_intercept=True)
model.fit(X = data.loc[:,["spy_excess","call_factor","put_factor"]],y = data["net_excess"])
print(f"Model betas: {model.coef_}")
print(f"Model intercept: {model.intercept_}")
print(f"Model R2: {r2_score(y_true = data['net_excess'],y_pred = model.predict(data.loc[:,['spy_excess','call_factor','put_factor']]))}")

pmh.calc_iterative_regression(multiple_y=data.loc[:,["net_excess"]],X=data.loc[:,["spy_excess","call_factor","put_factor"]],annual_factor=12)

Model betas: [ 0.47286157 -0.78970714  1.30343826]
Model intercept: 0.008456210729104473
Model R2: 0.056620461348179774
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Alpha  Annualized Alpha  R-Squared  spy_excess Beta  \
net_excess 0.0085            0.1015     0.0566           0.4729   

            call_factor Beta  put_factor Beta  Information Ratio  \
net_excess           -0.7897           1.3034             0.2697   

            Annualized Information Ratio  Tracking Error  \
net_excess                        0.9344          0.0313   

            Annualized Tracking Error  Fitted Mean  Annualized Fitted Mean  
net_excess                     0.1086       0.0131                  0.1566

In [77]:
# Fx carry trades
data_fx = pmh.read_excel_default(excel_name="./data/fx_rf_data.xlsx",sheet_name=1)
data_rf = pmh.read_excel_default(excel_name="./data/fx_rf_data.xlsx",sheet_name=0).apply(lambda x: x/252)
display(data_fx.head())
display(data_rf.head())
data_fx.index = pd.to_datetime(data_fx.index)
data_rf.index = pd.to_datetime(data_rf.index)
log_spot_data = np.log(data_fx)
log_rf_rate = np.log(1+data_rf)

tickers = data_fx.columns
results = {}
for ticker in tickers:
    X = pd.DataFrame({"log_spot_t_1": log_spot_data[ticker].shift(-1),
                      "log_spot_t_0": log_spot_data[ticker],
                      "log_rf_home": log_rf_rate[ticker],
                      "log_rf_foreign": log_rf_rate["USD"]}).dropna()
    y = X["log_spot_t_1"] - X["log_spot_t_0"] + X["log_rf_home"] - X["log_rf_foreign"]
    results[ticker] = y.to_frame(name = ticker)
results_df = pd.concat(results.values(),axis=1)
display(results_df.head())
display(pmh.calc_summary_statistics(results_df,annual_factor=252,provided_excess_returns=True).T)

JPY    EUR    GBP    MXN    CHF
date                                         
2008-01-01 0.0090 1.4592 1.9864 0.0918 0.8824
2008-01-02 0.0091 1.4715 1.9808 0.0916 0.8933
2008-01-03 0.0091 1.4750 1.9710 0.0919 0.9002
2008-01-04 0.0092 1.4743 1.9740 0.0915 0.9022
2008-01-07 0.0092 1.4696 1.9704 0.0918 0.8957

USD    JPY    EUR    GBP    MXN    CHF
date                                                
2008-01-01 0.0002 0.0000 0.0002 0.0002 0.0003 0.0001
2008-01-02 0.0002 0.0000 0.0001 0.0002 0.0003 0.0001
2008-01-03 0.0002 0.0000 0.0002 0.0002 0.0003 0.0001
2008-01-04 0.0002 0.0000 0.0002 0.0002 0.0003 0.0001
2008-01-07 0.0002 0.0000 0.0002 0.0002 0.0003 0.0001

JPY     EUR     GBP     MXN     CHF
date                                              
2008-01-01  0.0178  0.0084 -0.0028 -0.0022  0.0122
2008-01-02  0.0030  0.0024 -0.0049  0.0029  0.0076
2008-01-03  0.0065 -0.0005  0.0016 -0.0035  0.0021
2008-01-04 -0.0056 -0.0032 -0.0018  0.0025 -0.0073
2008-01-07  0.0025  0.0007  0.0015 -0.0026  0.0035

JPY                  EUR  \
Mean                                            -0.0001              -0.0001   
Annualized Mean                                 -0.0296              -0.0236   
Vol                                              0.0062               0.0057   
Annualized Vol                                   0.0991               0.0901   
Sharpe                                          -0.0188              -0.0165   
Annualized Sharpe                               -0.2982              -0.2623   
Min                                             -0.0548              -0.0243   
Max                                              0.0384               0.0345   
Skewness                                         0.1157               0.1087   
Excess Kurtosis                                  5.3461               2.3563   
Historical VaR (5.00%)                          -0.0096              -0.0090   
Annualized Historical VaR (5.00%)               -0.1518              -0.1422   
Historical CVaR (5.00%)                         -0.0142              -0.0129   
Annualized Historical CVaR (5.00%)              -0.2260              -0.2051   
Max Drawdown                                    -0.6243              -0.4820   
Peak                                2011-10-27 00:00:00  2008-07-10 00:00:00   
Bottom                              2024-07-09 00:00:00  2022-09-26 00:00:00   
Tangency Weights                                 0.7843               1.1924   
Recovery                                            NaT                  NaT   
Duration (days)                                     NaN                  NaN   
JPY Correlation                                  1.0000               0.2739   
EUR Correlation                                  0.2739               1.0000   
GBP Correlation                                  0.1490               0.6120   
MXN Correlation                                 -0.0830               0.3520   
CHF Correlation                                  0.3878               0.6485   

                                                    GBP                  MXN  \
Mean                                            -0.0001               0.0001   
Annualized Mean                                 -0.0248               0.0131   
Vol                                              0.0060               0.0082   
Annualized Vol                                   0.0957               0.1300   
Sharpe                                          -0.0164               0.0063   
Annualized Sharpe                               -0.2597               0.1007   
Min                                             -0.0840              -0.0793   
Max                                              0.0310               0.0668   
Skewness                                        -0.8302              -0.7031   
Excess Kurtosis                                 11.2960               7.7828   
Historical VaR (5.00%)                          -0.0093              -0.0127   
Annualized Historical VaR (5.00%)               -0.1478              -0.2017   
Historical CVaR (5.00%)                         -0.0140              -0.0196   
Annualized Historical CVaR (5.00%)              -0.2226              -0.3117   
Max Drawdown                                    -0.5095              -0.4093   
Peak                                2008-03-12 00:00:00  2008-08-01 00:00:00   
Bottom                              2022-09-23 00:00:00  2017-01-18 00:00:00   
Tangency Weights                                 0.5998              -0.4564   
Recovery                                            NaT  2023-05-09 00:00:00   
Duration (days)                                     NaN           2,302.0000   
JPY Correlation                                  0.1490              -0.0830   
EUR Correlation                                  0.6120               0.3520   
GBP Correlation                                  1.0000               0.3673   
MXN Correlation                                  0.

In [102]:
# predicting fx
results = {}
betas = {}
alphas = {}
for ticker in tickers:
    main = pd.DataFrame({"log_spot_t_1": log_spot_data[ticker].shift(-1),
                      "log_spot_t_0": log_spot_data[ticker],
                      "log_rf_home": log_rf_rate[ticker],
                      "log_rf_foreign": log_rf_rate["USD"]}).dropna()
    y = main["log_spot_t_1"] - main["log_spot_t_0"] 
    X = main["log_rf_foreign"] - main["log_rf_home"]
    model_fit = pmh.calc_iterative_regression(multiple_y=y.to_frame(f"{ticker}_spot_spread"),X=X.to_frame(f"{ticker}_rf_spread"),annual_factor=252)
    results[ticker] = model_fit.T
    betas[ticker] = model_fit.T.loc[f"{ticker}_rf_spread Beta"]
    alphas[ticker] = model_fit.T.loc["Annualized Alpha"]
    display(model_fit.T)



"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


JPY_spot_spread
Alpha                                 -0.0001
Annualized Alpha                      -0.0219
R-Squared                              0.0000
JPY_rf_spread Beta                     0.3518
Treynor Ratio                         -0.0002
Annualized Treynor Ratio              -0.0504
Information Ratio                     -0.0139
Annualized Information Ratio          -0.2207
Tracking Error                         0.0062
Annualized Tracking Error              0.0991
Fitted Mean                           -0.0001
Annualized Fitted Mean                -0.0177

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


EUR_spot_spread
Alpha                                 -0.0001
Annualized Alpha                      -0.0315
R-Squared                              0.0003
EUR_rf_spread Beta                     2.1602
Treynor Ratio                         -0.0000
Annualized Treynor Ratio              -0.0078
Information Ratio                     -0.0220
Annualized Information Ratio          -0.3500
Tracking Error                         0.0057
Annualized Tracking Error              0.0900
Fitted Mean                           -0.0001
Annualized Fitted Mean                -0.0168

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


GBP_spot_spread
Alpha                                 -0.0001
Annualized Alpha                      -0.0252
R-Squared                              0.0015
GBP_rf_spread Beta                     6.8103
Treynor Ratio                         -0.0000
Annualized Treynor Ratio              -0.0036
Information Ratio                     -0.0166
Annualized Information Ratio          -0.2639
Tracking Error                         0.0060
Annualized Tracking Error              0.0956
Fitted Mean                           -0.0001
Annualized Fitted Mean                -0.0248

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


MXN_spot_spread
Alpha                                 -0.0003
Annualized Alpha                      -0.0736
R-Squared                              0.0000
MXN_rf_spread Beta                    -0.8088
Treynor Ratio                          0.0002
Annualized Treynor Ratio               0.0430
Information Ratio                     -0.0357
Annualized Information Ratio          -0.5660
Tracking Error                         0.0082
Annualized Tracking Error              0.1300
Fitted Mean                           -0.0001
Annualized Fitted Mean                -0.0348

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


CHF_spot_spread
Alpha                                  0.0000
Annualized Alpha                       0.0037
R-Squared                              0.0001
CHF_rf_spread Beta                     0.9069
Treynor Ratio                          0.0001
Annualized Treynor Ratio               0.0172
Information Ratio                      0.0022
Annualized Information Ratio           0.0346
Tracking Error                         0.0067
Annualized Tracking Error              0.1069
Fitted Mean                            0.0001
Annualized Fitted Mean                 0.0156

In [113]:
# Dynamic carry trade
forecasts = {}
for ticker in tickers:
    main = pd.DataFrame({"log_spot_t_1": log_spot_data[ticker].shift(-1),
                      "log_spot_t_0": log_spot_data[ticker],
                      "log_rf_home": log_rf_rate[ticker],
                      "log_rf_foreign": log_rf_rate["USD"]}).dropna()
    X = (main["log_rf_foreign"] - main["log_rf_home"]).dropna()
    beta = betas[ticker].values
    alpha = (alphas[ticker].values)/252
    print(alpha)
    print(beta)
    forecast = alpha + (beta-1)*X

    forecasts[ticker] = forecast.to_frame(name = ticker)
forecasts_df = pd.concat(forecasts.values(),axis=1)
display(forecasts_df.head())
print(betas)
print(alphas)
from itables import show
show(forecasts_df)

print(forecasts_df.apply(lambda x: x>0).sum() / len(forecasts_df))
    

[-8.68154122e-05]
[0.35175526]
[-0.00012505]
[2.16019188]
[-0.00010011]
[6.81027245]
[-0.00029191]
[-0.80876561]
[1.46942146e-05]
[0.90694452]


JPY     EUR     GBP     MXN    CHF
date                                             
2008-01-01 -0.0002 -0.0001 -0.0004 -0.0001 0.0000
2008-01-02 -0.0002 -0.0001 -0.0004 -0.0001 0.0000
2008-01-03 -0.0002 -0.0001 -0.0004 -0.0001 0.0000
2008-01-04 -0.0002 -0.0001 -0.0004 -0.0001 0.0000
2008-01-07 -0.0002 -0.0001 -0.0004 -0.0001 0.0000

{'JPY': JPY_spot_spread   0.3518
Name: JPY_rf_spread Beta, dtype: float64, 'EUR': EUR_spot_spread   2.1602
Name: EUR_rf_spread Beta, dtype: float64, 'GBP': GBP_spot_spread   6.8103
Name: GBP_rf_spread Beta, dtype: float64, 'MXN': MXN_spot_spread   -0.8088
Name: MXN_rf_spread Beta, dtype: float64, 'CHF': CHF_spot_spread   0.9069
Name: CHF_rf_spread Beta, dtype: float64}
{'JPY': JPY_spot_spread   -0.0219
Name: Annualized Alpha, dtype: float64, 'EUR': EUR_spot_spread   -0.0315
Name: Annualized Alpha, dtype: float64, 'GBP': GBP_spot_spread   -0.0252
Name: Annualized Alpha, dtype: float64, 'MXN': MXN_spot_spread   -0.0736
Name: Annualized Alpha, dtype: float64, 'CHF': CHF_spot_spread   0.0037
Name: Annualized Alpha, dtype: float64}


JPY   0.0000
EUR   0.0378
GBP   0.2489
MXN   0.7602
CHF   0.9652
dtype: float64


In [64]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import os
import sys
from sklearn.linear_model import LinearRegression
from itables import show

notebook = "homework"
if notebook in os.getcwd():
    os.chdir(os.pardir)
    print(f"Changed working directory to {os.getcwd()}")


In [65]:

def load_data(data_path: str, filename: str, sheet_name: str):
    full_data_path = os.path.join(data_path, filename)
    data = pd.read_excel(full_data_path, sheet_name=sheet_name, parse_dates=True, skiprows=2, index_col=0)
    data = data.dropna()
    if "date" in data.columns:
        data = data.set_index("date")
    elif "Date" in data.columns:
        data = data.set_index("Date")
    return data

# Load LTCM and GMO data
data_path = "./data"
data_2 = load_data(data_path, "ltcm_exhibits_data.xlsx", sheet_name="Exhibit 2")
data_4 = load_data(data_path, "ltcm_exhibits_data.xlsx", sheet_name="Exhibit 4")

def load_gmo_data(data_path: str, filename: str, sheet_name: str):
    full_data_path = os.path.join(data_path, filename)
    data = pd.read_excel(full_data_path, sheet_name=sheet_name, parse_dates=True, index_col=0)
    data = data.dropna()
    return data

spy_data = load_gmo_data(data_path, "gmo_analysis_data.xlsx", sheet_name="total returns")
rf_data = load_gmo_data(data_path, "gmo_analysis_data.xlsx", sheet_name="risk-free rate")

# Align indices and convert to period-based monthly data
data_2.index = pd.to_datetime(data_2.index).to_period("M")
data_4.index = pd.to_datetime(data_4.index).to_period("M")
spy_data.index = pd.to_datetime(spy_data.index).to_period("M")
rf_data.index = pd.to_datetime(rf_data.index).to_period("M")

# Adjust risk-free rate for alignment
rf_data["TBill 3M"] = rf_data["TBill 3M"].shift(1)
spy_data["SPY"] = spy_data["SPY"].shift(1)
# Merge SPY and Risk-Free data
spy_excess_data = pd.merge(spy_data, rf_data, left_index=True, right_index=True, how="left").dropna()
spy_excess_data["SPY_Excess"] = spy_excess_data["SPY"] - spy_excess_data["TBill 3M"]

# Merge LTCM data with Risk-Free data
data_2_excess = pd.merge(data_2, rf_data, left_index=True, right_index=True, how="left").dropna()
data_2_excess["excess_gross"] = data_2_excess["Gross Monthly Performancea"] - data_2_excess["TBill 3M"]
data_2_excess["excess_net"] = data_2_excess["Net Monthly Performanceb"] - data_2_excess["TBill 3M"]
data_2_excess = data_2_excess.loc[:, ["excess_gross", "excess_net"]]

# Function to calculate statistics
def calculate_stats(data: pd.DataFrame, annual_factor: int = 12, quantile: float = 0.05):
    if isinstance(data, pd.Series):
        data = data.to_frame()
    stats = pd.DataFrame(index=data.columns)
    stats["skew"] = data.skew()
    stats["kurtosis"] = data.kurtosis()
    stats["mean"] = data.mean()
    stats["std"] = data.std()
    stats["sharpe"] = stats["mean"] / stats["std"]
    stats["annualized_mean"] = stats["mean"] * annual_factor
    stats["annualized_std"] = stats["std"] * np.sqrt(annual_factor)
    stats["annualized_sharpe"] = stats["sharpe"] * np.sqrt(annual_factor)
    stats["VaR"] = data.quantile(quantile)
    stats["CVaR"] = data[data < data.quantile(quantile)].mean()
    drawdown = data.apply(lambda x: (1 + x).cumprod() / (1 + x).cumprod().cummax() - 1)
    stats["max_drawdown"] = drawdown.min()
    return stats

# Calculate statistics for LTCM and SPY data
ltcma_stats = calculate_stats(data_2_excess)
spy_stats = calculate_stats(spy_excess_data.loc[data_2_excess.first_valid_index():data_2_excess.last_valid_index(),"SPY_Excess"])

# Display results
display(ltcma_stats)
display(spy_stats)

skew  kurtosis      mean       std    sharpe  \
excess_gross -0.208351  1.772578 -0.025785  0.039182 -0.658083   
excess_net   -0.732503  3.110232 -0.033011  0.032127 -1.027535   

              annualized_mean  annualized_std  annualized_sharpe      VaR  \
excess_gross        -0.309419         0.13573          -2.279666 -0.07254   
excess_net          -0.396136         0.11129          -3.559485 -0.06920   

                  CVaR  max_drawdown  
excess_gross -0.116067     -0.749191  
excess_net   -0.111733     -0.827814

skew  kurtosis      mean       std    sharpe  annualized_mean  \
SPY_Excess -0.521066 -0.942425 -0.026968  0.040551 -0.665023        -0.323611   

            annualized_std  annualized_sharpe       VaR      CVaR  \
SPY_Excess        0.140474          -2.303706 -0.097719 -0.101303   

            max_drawdown  
SPY_Excess     -0.375284

LTCM's gross and net excess returns demonstrate significantly worse performance compared to SPY. The annualized mean return for LTCM was deeply negative, with net returns (-39.61%) underperforming gross returns (-30.94%) due to the impact of fees. While LTCM’s strategies exhibited lower volatility than SPY (13.57% gross, 11.13% net vs. 14.05% for SPY), its Sharpe ratios were abysmal, particularly for net returns (-3.56 vs. -2.28 gross and -2.30 for SPY). Tail risk metrics highlight greater left-tail vulnerability for LTCM net returns, with higher negative skewness (-0.73) and kurtosis (3.11), compared to SPY (-0.52 skew, -0.94 kurtosis). Although LTCM’s VaR at the 5% level (-7.25% gross, -6.92% net) suggests slightly smaller extreme losses than SPY (-9.77%), LTCM’s poor mean returns and amplified fees make its overall performance significantly worse. Net results indicate that fees not only reduced returns but also exacerbated LTCM's inability to provide risk-adjusted outperformance relative to SPY.

## Question 2.2: Regressions


In [66]:
full_data = pd.merge(data_2_excess, spy_excess_data, left_index=True, right_index=True, how="inner")
display(full_data)


excess_gross  excess_net       SPY     GMWAX     GMGEX  TBill 3M  \
1997-01        0.0107     -0.0083 -0.023292  0.014735  0.034448    0.0503   
1997-02       -0.0441     -0.0451  0.061786  0.022265  0.012733    0.0501   
1997-03       -0.0304     -0.0354  0.009565 -0.015152 -0.016441    0.0514   
1997-04       -0.0336     -0.0416 -0.045721 -0.006731  0.000000    0.0516   
1997-05       -0.0085     -0.0185  0.064368  0.057115  0.058014    0.0505   
1997-06       -0.0673     -0.0633  0.063207  0.043040  0.044610    0.0493   
1997-07       -0.0325     -0.0405  0.041116  0.029851  0.037367    0.0505   
1997-08       -0.0434     -0.0444  0.079264 -0.034101 -0.023156    0.0514   
1997-09       -0.0045     -0.0145 -0.051803  0.048544  0.055312    0.0495   
1997-10       -0.0407     -0.0467  0.048088 -0.076599 -0.064060    0.0497   
1997-11       -0.0614     -0.0594 -0.024503 -0.000912  0.015111    0.0514   
1997-12       -0.0206     -0.0276  0.038697  0.013686 -0.084841    0.0516   
1998-01       -0.0634     -0.0684  0.019097 -0.000900  0.001063    0.0504   
1998-02       -0.0389     -0.0389  0.012878  0.060661  0.058624    0.0509   
1998-03       -0.0533     -0.0533  0.069295  0.054077  0.043038    0.0503   
1998-04       -0.0225     -0.0305  0.048761 -0.006446 -0.005482    0.0495   
1998-05       -0.1170     -0.1140  0.012791 -0.018383 -0.014700    0.0500   
1998-06       -0.1508     -0.1508 -0.020769 -0.030019 -0.008441    0.0498   
1998-07       -0.0446     -0.0496  0.042590 -0.000284 -0.033756    0.0496   

         SPY_Excess  
1997-01   -0.073592  
1997-02    0.011686  
1997-03   -0.041835  
1997-04   -0.097321  
1997-05    0.013868  
1997-06    0.013907  
1997-07   -0.009384  
1997-08    0.027864  
1997-09   -0.101303  
1997-10   -0.001612  
1997-11   -0.075903  
1997-12   -0.012903  
1998-01   -0.031303  
1998-02   -0.038022  
1998-03    0.018995  
1998-04   -0.000739  
1998-05   -0.037209  
1998-06   -0.070569  
1998-07   -0.007010

In [67]:
X = full_data["SPY_Excess"].values.reshape(-1, 1)
y = full_data["excess_net"].values.reshape(-1, 1)

reg = LinearRegression(fit_intercept=True).fit(X, y)
print(f"R^2: {reg.score(X, y)}")
print(f"Intercept: {reg.intercept_[0]}")
print(f"Coefficient: {reg.coef_[0][0]}")


R^2: 0.00841421440712986
Intercept: -0.047993096762149655
Coefficient: 0.0759803569348601


b. From the regression LTCM, LTCM does not appear to be a closet indexer. Its strategy seems to involve idiosyncratic, market-neutral trades rather than replicating market index movements. A closet indexer is a fund that claims to follow active management but closely tracks a market index, reflected by high beta high R2 (large proportion of returns explained by the market)

c. α=−0.048: LTCM delivers negative independent excess returns on average, meaning it underperforms even after accounting for market exposure. LTCM does not deliver excess returns beyond the risk premium expected from market exposure. In fact, its negative alpha suggests poor performance even in isolation from the market.


## Question 2.3

In [68]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline, make_pipeline
model = LinearRegression(fit_intercept=True)
poly = PolynomialFeatures(degree=2, include_bias=False)
pipe = Pipeline([("Polytransform", poly), ("LinearModel", model)])
pipe.fit(X, y)
print(f"R^2: {pipe.score(X, y)}")
print(f"Intercept: {pipe.named_steps['LinearModel'].intercept_[0]}")
print(f"Coefficients: {pipe.named_steps['LinearModel'].coef_[0]}")
coefficient_data = pd.DataFrame({"Feature": pipe[:-1].get_feature_names_out(), "Coefficient": pipe.named_steps["LinearModel"].coef_[0]}).set_index("Feature").T
display(coefficient_data)


R^2: 0.08038651929099849
Intercept: -0.050066719296108515
Coefficients: [0.54685615 6.46443459]


Feature            x0      x0^2
Coefficient  0.546856  6.464435

b. In the earlier linear regression only 0.84% of LTCM’s variation was explained by the market. Adding the quadratic term increases R2 to 8%. This is a significant improvement, indicating that LTCM’s returns have a nonlinear relationship with market returns.

The quadratic market factor substantially increases the explained variation in LTCM’s returns, suggesting that LTCM’s strategy exhibits nonlinear sensitivity to market returns.

c. β2 =6.4644 is positive, meaning LTCM's returns increase disproportionately when the market experiences large absolute returns (either positive or negative).
This is consistent with being long market options (both calls and puts), as such a position benefits from increased market volatility, regardless of the direction of the market. As volatility increases, the uncertainty about the underlying asset's future price grows. This increases the likelihood that the asset price will significantly exceed the strike price (for a call) or fall below the strike price (for a put), increasing the potential payoff of the option. Hence, option values rise with higher volatility.

d. The positive polynomial term beta indicates that LTCM’s returns improve as market volatility increases (larger squared returns). This suggests that LTCM is positively exposed to market volatility, as its performance improves during periods of significant market movement.

## Question 2.4

In [69]:
# up_market factor and down market factor

data_up_down = full_data.copy()
one_std_dev = 0.03
data_up_down["up_market"] = np.where((data_up_down["SPY_Excess"] - one_std_dev) > 0,(data_up_down["SPY_Excess"] - one_std_dev),0 )
data_up_down["down_market"] = np.where((-one_std_dev-data_up_down["SPY_Excess"] ) > 0,(one_std_dev-data_up_down["SPY_Excess"] ),0 )
model = LinearRegression(fit_intercept=True)
X = data_up_down[["SPY_Excess","up_market", "down_market"]].values
y = data_up_down["excess_net"].values
model.fit(X, y)
print(f"R^2: {model.score(X, y)}")
print(f"Intercept: {model.intercept_}")
print(f"Coefficients: {model.coef_}")



R^2: 0.06699747929994015
Intercept: -0.04104811434881364
Coefficients: [-0.6654421   0.         -0.61148545]


b. LTCM shows no sensitivity to the call-like factor, meaning it neither benefits from nor loses to large upward market movements. LTCM is effectively neutral on the call-like factor. Put-Like Factor. The negative indicates that LTCM is short the put-like factor, implying significant losses when markets experience large downward movements beyond the threshold.

c. Put like factor.  LTCM’s sensitivity to market movements arises primarily from exposure to large downward market movements.

d. LTCM shows no exposure to large upward market movements, indicating it does not gain from increased market volatility on the upside. LTCM is negatively exposed to large downward market movements, suggesting its volatility exposure comes primarily from the downside. Overall, since LTCM is short the downside and has no exposure to the upside, its sensitivity to market volatility is primarily driven by its negative exposure to downward market movements.

## Question 3: FX Carry Trade

In [70]:
fx_data_rf = pd.read_excel("./data/fx_rf_data.xlsx",sheet_name = 0,parse_dates=True,index_col=0)
fx_data_exchange = pd.read_excel("./data/fx_rf_data.xlsx",sheet_name = 1,parse_dates=True,index_col=0)

display(fx_data_rf)
display(fx_data_exchange)

USD      JPY       EUR       GBP    MXN       CHF
date                                                              
2008-01-01  0.044000  0.00478  0.042500  0.057000  0.075  0.020645
2008-01-02  0.041563  0.00478  0.036000  0.054750  0.075  0.020645
2008-01-03  0.043463  0.00478  0.039400  0.056000  0.075  0.021500
2008-01-04  0.042188  0.00478  0.040650  0.055000  0.075  0.020950
2008-01-07  0.042188  0.00493  0.041000  0.054600  0.075  0.020200
...              ...      ...       ...       ...    ...       ...
2024-10-25  0.048700  0.00227  0.032000  0.049575  0.105  0.008450
2024-10-28  0.048650  0.00228  0.032088  0.050500  0.105  0.008400
2024-10-29  0.048625  0.00228  0.031750  0.050500  0.105  0.008350
2024-10-30  0.048625  0.00227  0.032275  0.050500  0.105  0.008350
2024-10-31  0.049275  0.00227  0.031500  0.049575  0.105  0.008200

[4393 rows x 6 columns]

JPY     EUR     GBP      MXN     CHF
date                                                 
2008-01-01  0.008957  1.4592  1.9864  0.09183  0.8824
2008-01-02  0.009119  1.4715  1.9808  0.09162  0.8933
2008-01-03  0.009148  1.4750  1.9710  0.09187  0.9002
2008-01-04  0.009209  1.4743  1.9740  0.09154  0.9022
2008-01-07  0.009159  1.4696  1.9704  0.09176  0.8957
...              ...     ...     ...      ...     ...
2024-10-25  0.006565  1.0796  1.2962  0.05003  1.1536
2024-10-28  0.006524  1.0812  1.2972  0.04989  1.1555
2024-10-29  0.006521  1.0819  1.3015  0.04984  1.1530
2024-10-30  0.006518  1.0856  1.2962  0.04962  1.1538
2024-10-31  0.006578  1.0884  1.2899  0.04991  1.1573

[4393 rows x 5 columns]

In [71]:
# Data processing
fx_data_rf_logs = fx_data_rf.apply(np.log1p, axis = 0)
fx_data_exchange_logs = fx_data_exchange.apply(np.log, axis = 0)
display(fx_data_rf_logs)
display(fx_data_exchange_logs)

USD       JPY       EUR       GBP       MXN       CHF
date                                                                  
2008-01-01  0.043059  0.004769  0.041622  0.055435  0.072321  0.020435
2008-01-02  0.040722  0.004769  0.035367  0.053304  0.072321  0.020435
2008-01-03  0.042545  0.004769  0.038644  0.054488  0.072321  0.021272
2008-01-04  0.041322  0.004769  0.039846  0.053541  0.072321  0.020734
2008-01-07  0.041322  0.004918  0.040182  0.053162  0.072321  0.019999
...              ...       ...       ...       ...       ...       ...
2024-10-25  0.047551  0.002267  0.031499  0.048385  0.099845  0.008414
2024-10-28  0.047504  0.002277  0.031584  0.049266  0.099845  0.008365
2024-10-29  0.047480  0.002277  0.031256  0.049266  0.099845  0.008315
2024-10-30  0.047480  0.002267  0.031765  0.049266  0.099845  0.008315
2024-10-31  0.048099  0.002267  0.031014  0.048385  0.099845  0.008167

[4393 rows x 6 columns]

JPY       EUR       GBP       MXN       CHF
date                                                        
2008-01-01 -4.715320  0.377888  0.686324 -2.387816 -0.125110
2008-01-02 -4.697395  0.386282  0.683501 -2.390106 -0.112833
2008-01-03 -4.694220  0.388658  0.678541 -2.387381 -0.105138
2008-01-04 -4.687574  0.388183  0.680062 -2.390979 -0.102919
2008-01-07 -4.693018  0.384990  0.678237 -2.388579 -0.110150
...              ...       ...       ...       ...       ...
2024-10-25 -5.026003  0.076591  0.259437 -2.995132  0.142887
2024-10-28 -5.032268  0.078072  0.260208 -2.997935  0.144533
2024-10-29 -5.032728  0.078719  0.263517 -2.998937  0.142367
2024-10-30 -5.033188  0.082133  0.259437 -3.003361  0.143061
2024-10-31 -5.024025  0.084709  0.254565 -2.997534  0.146090

[4393 rows x 5 columns]

In [72]:
full_fx_data = pd.merge(fx_data_rf_logs, fx_data_exchange_logs.rename(columns = lambda x: f"{x}_spot"), left_index=True, right_index=True, how="inner")

list_targets = ["JPY","EUR","GBP","MXN","CHF"]
usd_col = "USD"
for target in list_targets:
    target_col = target
    full_fx_data[f"{target_col}_excess_relative_to_usd"] = full_fx_data[f"{target_col}_spot"] -  full_fx_data[f"{target_col}_spot"].shift(1) + full_fx_data[target_col].shift(1) - full_fx_data[usd_col].shift(1)
    display(full_fx_data[[f"{target_col}_excess_relative_to_usd"]].head())
stats_df = calculate_stats(full_fx_data.loc[:,[col for col in full_fx_data.columns if "excess_relative_to_usd" in col]].dropna())
display(stats_df)



JPY_excess_relative_to_usd
date                                  
2008-01-01                         NaN
2008-01-02                   -0.020366
2008-01-03                   -0.032779
2008-01-04                   -0.031130
2008-01-07                   -0.041998

EUR_excess_relative_to_usd
date                                  
2008-01-01                         NaN
2008-01-02                    0.006956
2008-01-03                   -0.002980
2008-01-04                   -0.004376
2008-01-07                   -0.004670

GBP_excess_relative_to_usd
date                                  
2008-01-01                         NaN
2008-01-02                    0.009552
2008-01-03                    0.007622
2008-01-04                    0.013464
2008-01-07                    0.010393

MXN_excess_relative_to_usd
date                                  
2008-01-01                         NaN
2008-01-02                    0.026972
2008-01-03                    0.034323
2008-01-04                    0.026177
2008-01-07                    0.033399

CHF_excess_relative_to_usd
date                                  
2008-01-01                         NaN
2008-01-02                   -0.010348
2008-01-03                   -0.012593
2008-01-04                   -0.019054
2008-01-07                   -0.027819

skew  kurtosis      mean       std    sharpe  \
JPY_excess_relative_to_usd -1.242156  0.918384 -0.011690  0.017030 -0.686440   
EUR_excess_relative_to_usd  0.288571  0.473907 -0.006780  0.012497 -0.542496   
GBP_excess_relative_to_usd  0.432617  3.115715 -0.000171  0.010263 -0.016640   
MXN_excess_relative_to_usd  0.062165  1.223848  0.045972  0.013329  3.448995   
CHF_excess_relative_to_usd  0.197128  8.512459 -0.012832  0.014569 -0.880780   

                            annualized_mean  annualized_std  \
JPY_excess_relative_to_usd        -0.140284        0.058995   
EUR_excess_relative_to_usd        -0.081357        0.043292   
GBP_excess_relative_to_usd        -0.002049        0.035551   
MXN_excess_relative_to_usd         0.551664        0.046173   
CHF_excess_relative_to_usd        -0.153982        0.050468   

                            annualized_sharpe       VaR      CVaR  \
JPY_excess_relative_to_usd          -2.377898 -0.051958 -0.055834   
EUR_excess_relative_to_usd          -1.879260 -0.027230 -0.030475   
GBP_excess_relative_to_usd          -0.057642 -0.016861 -0.020924   
MXN_excess_relative_to_usd          11.947668  0.024438  0.019163   
CHF_excess_relative_to_usd          -3.051111 -0.038334 -0.042230   

                            max_drawdown  
JPY_excess_relative_to_usd     -1.000000  
EUR_excess_relative_to_usd     -1.000000  
GBP_excess_relative_to_usd     -0.999996  
MXN_excess_relative_to_usd     -0.038885  
CHF_excess_relative_to_usd     -1.000000

Based on the results, the mean excess log returns are mixed across currencies. For example, the MXN has a high positive mean return of 0.045977 (5.52% annualized), indicating strong excess returns relative to USD, while others, like the JPY and CHF, show negative mean returns, reflecting poor performance relative to the USD. The volatility varies, with GBP showing the lowest volatility (0.01028, annualized 3.56%) and MXN the highest (0.013305, annualized 4.61%), indicating greater risk associated with holding MXN relative to USD.

The Sharpe ratios highlight significant differences in risk-adjusted performance. The MXN has an exceptionally high Sharpe ratio of 3.46, suggesting strong excess returns relative to its risk. In contrast, the other currencies, particularly JPY and CHF, exhibit negative Sharpe ratios, reflecting underperformance relative to USD even after adjusting for volatility. This divergence underscores the impact of different macroeconomic conditions, interest rate policies, and currency stability across regions.

## Question 3.2: Implications of UIP

1. Yes the stats contradict the UIP which states that by expectation, there should not be any mean excess log return, as this would entail arbitrage. Any premium from interest rate differentials will be reflected in the expected spot rate changes, meaning excess log returns should average to 0 over long period. MXN has high positive mean excess returns, JPY and CHF show negative mean excess returns.

2. The MXN offered the best Sharpe ratio of 3.46, indicating exceptionally high risk-adjusted returns. This reflects strong mean excess returns coupled with relatively low volatility compared to the returns, making MXN the most profitable currency to hold during the sample period.

3. Yes, long positions in the JPY (−0.011692 annualized -1.40%) and CHF ( −0.012835, annualized -1.54%) earned negative excess returns relative to the USD. These results suggest that holding these currencies during the sample period led to consistent losses when compared to USD-denominated investments.

## Question 3.3: Predicting FX

In [73]:
predict_fx_data = fx_data_rf.copy()
predict_fx_data = pd.concat([predict_fx_data,full_fx_data.loc[:,[col for col in full_fx_data.columns if "spot" in col]]],axis=1)
predict_fx_data = predict_fx_data.dropna()
# Predicting FX rates
usd_col = "USD"
results = {}
for target in list_targets:
    target_col = target
    X = (predict_fx_data.loc[:,usd_col] - predict_fx_data.loc[:,target_col]).values.reshape(-1, 1)[:-1]
    y = (predict_fx_data[f"{target_col}_spot"].shift(-1)  - predict_fx_data[f"{target_col}_spot"]).dropna().values
    model = LinearRegression(fit_intercept=True)
    model.fit(X, y)
    results[target] = {"model": model, "R^2": model.score(X, y), "intercept": model.intercept_, "beta": model.coef_[0]}

results_df = pd.DataFrame(results).T
display(results_df)

model       R^2 intercept      beta
JPY  LinearRegression()  0.000013 -0.000087  0.001395
EUR  LinearRegression()  0.000298 -0.000125  0.008572
GBP  LinearRegression()  0.001537   -0.0001  0.027023
MXN  LinearRegression()   0.00002  -0.00029 -0.003148
CHF  LinearRegression()   0.00005  0.000015  0.003604

If the foreign risk-free rate increases relative to the US risk-free rate, the interest rate differential widens, theoretically making foreign currency assets more attractive due to higher returns. According to Uncovered Interest Parity (UIP), this should lead to an appreciation of the foreign currency (spot rate increases), resulting in USD depreciation. A negative 𝛽 indicates a predicted strengthening of the USD relative to the foreign currency, as the foreign currency is expected to weaken when the interest rate differential widens. From the results:

MXN  predicts a relative strengthening of the USD, meaning the foreign currency (MXN) would depreciate.
This contradicts UIP, as an increase in foreign interest rates should typically lead to foreign currency appreciation.

A positive β indicates a predicted weakening of the USD relative to the foreign currency, as the foreign currency is expected to appreciate when the interest rate differential widens. From the results all except MXN  align with UIP predictions, where an increase in foreign risk-free rates is associated with foreign currency appreciation and USD depreciation.

FX predictability is strongest for the currency with the highest β in magnitude, as it indicates greater sensitivity of the spot rate to changes in the risk-free rate differential.

GBP has the largest β=0.027023, suggesting that the GBP spot rate is the most responsive to changes in the interest rate differential. However, the R2 =0.001537 for GBP remains very low, indicating that this predictability is weak in practice and that most of the spot rate variation is driven by factors other than interest rate differentials.

## Question 3.4: Predicting FX

In [ ]:
# corrected beta 
betas = (results_df.loc[:, "beta"] - 1).to_frame()
alphas = results_df.loc[:, "intercept"].to_frame()
compiled_predictions = pd.DataFrame()
for target in list_targets:
    target_col = target
    # Calculate X as the interest rate differential
    X = (predict_fx_data.loc[:, usd_col] - predict_fx_data.loc[:, target_col]).values.reshape(-1, 1)[:-1]
    
    # Predict spot rate changes using corrected beta and alpha
    pred_y = (betas.loc[target, "beta"] * X + alphas.loc[target, "intercept"]).flatten()
    
    # Actual spot rate changes (aligned properly with predictions)
    y_actual = (predict_fx_data[f"{target_col}_spot"].shift(-1) - predict_fx_data[f"{target_col}_spot"]).dropna().values
    
    # Ensure predictions and actual values have the same length
    pred_y = pred_y[:len(y_actual)]
    
    # Create DataFrame with predictions and actuals
    pred_df = pd.DataFrame({f"pred_y_{target_col}": pred_y, f"y_{target_col}": y_actual})
    
    compiled_predictions = pd.concat([compiled_predictions, pred_df], axis=1)
compiled_predictions.set_index(predict_fx_data.index[:-1], inplace=True)

# Display final results
display(compiled_predictions)


pred_y_JPY     y_JPY  pred_y_EUR     y_EUR  pred_y_GBP     y_GBP  \
date                                                                           
2008-01-01   -0.039252  0.017925   -0.001612  0.008394    0.012549 -0.002823   
2008-01-02   -0.036818  0.003175   -0.005640  0.002376    0.012731 -0.004960   
2008-01-03   -0.038716  0.006646   -0.004153 -0.000475    0.012098  0.001521   
2008-01-04   -0.037443 -0.005444   -0.001650 -0.003193    0.012366 -0.001825   
2008-01-07   -0.037293  0.002617   -0.001303  0.000748    0.011976  0.001420   
...                ...       ...         ...       ...         ...       ...   
2024-10-24   -0.046402 -0.003194   -0.017128 -0.002960    0.000874 -0.001002   
2024-10-25   -0.046452 -0.006265   -0.016682  0.001481    0.000751  0.000771   
2024-10-28   -0.046392 -0.000460   -0.016545  0.000647    0.001700  0.003309   
2024-10-29   -0.046367 -0.000460   -0.016855  0.003414    0.001724 -0.004081   
2024-10-30   -0.046377  0.009163   -0.016335  0.002576    0.001724 -0.004872   

            pred_y_MXN     y_MXN  pred_y_CHF     y_CHF  
date                                                    
2008-01-01    0.030808 -0.002289   -0.023256  0.012277  
2008-01-02    0.033253  0.002725   -0.020828  0.007694  
2008-01-03    0.031347 -0.003598   -0.021869  0.002219  
2008-01-04    0.032626  0.002400   -0.021147 -0.007231  
2008-01-07    0.032626 -0.002728   -0.021894  0.003566  
...                ...       ...         ...       ...  
2024-10-24    0.056238 -0.008954   -0.039991 -0.001299  
2024-10-25    0.056188 -0.002802   -0.040090  0.001646  
2024-10-28    0.056238 -0.001003   -0.040090 -0.002166  
2024-10-29    0.056263 -0.004424   -0.040115  0.000694  
2024-10-30    0.056263  0.005827   -0.040115  0.003029  

[4391 rows x 10 columns]

In [81]:
compiled_predictions_subset = compiled_predictions.loc[:,[col for col in compiled_predictions.columns if "pred_y" in col]]
compiled_predictions_subset.columns = [col.replace("pred_y_", "") for col in compiled_predictions_subset.columns]
display(compiled_predictions_subset)
compiled_predictions_subset = compiled_predictions_subset.applymap(lambda x: 1 if x > 0 else 0)

# Frequency of correct predictions
correct_predictions = compiled_predictions_subset.sum() / len(compiled_predictions_subset)
display(correct_predictions)


JPY       EUR       GBP       MXN       CHF
date                                                        
2008-01-01 -0.039252 -0.001612  0.012549  0.030808 -0.023256
2008-01-02 -0.036818 -0.005640  0.012731  0.033253 -0.020828
2008-01-03 -0.038716 -0.004153  0.012098  0.031347 -0.021869
2008-01-04 -0.037443 -0.001650  0.012366  0.032626 -0.021147
2008-01-07 -0.037293 -0.001303  0.011976  0.032626 -0.021894
...              ...       ...       ...       ...       ...
2024-10-24 -0.046402 -0.017128  0.000874  0.056238 -0.039991
2024-10-25 -0.046452 -0.016682  0.000751  0.056188 -0.040090
2024-10-28 -0.046392 -0.016545  0.001700  0.056238 -0.040090
2024-10-29 -0.046367 -0.016855  0.001724  0.056263 -0.040115
2024-10-30 -0.046377 -0.016335  0.001724  0.056263 -0.040115

[4391 rows x 5 columns]

C:\Users\Andik\AppData\Local\Temp\ipykernel_24868\3527264759.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  compiled_predictions_subset = compiled_predictions_subset.applymap(lambda x: 1 if x > 0 else 0)


JPY    0.000911
EUR    0.259622
GBP    0.521066
MXN    1.000000
CHF    0.026418
dtype: float64

2. MXN almost always have a positve FX risk premium, JPY almost always go negative

3. The static carry trade involves borrowing in a low-interest-rate currency and investing in a high-interest-rate currency, assuming a consistent positive return based solely on the interest rate differential. However, this approach does not account for fluctuations in exchange rates, which can erode or reverse these returns. By incorporating conditional risk premia, the carry trade can be significantly improved.

The conditional approach leverages forecasts for the FX risk premium to determine when the carry trade is likely to generate positive returns. Specifically, trades are only executed when the expected FX risk premium is positive:This ensures that the trade not only benefits from favorable interest rate differentials but also aligns with predicted favorable exchange rate movements. For instance, if the interest rate differential suggests a profitable trade but the regression model predicts a negative risk premium due to anticipated adverse exchange rate movements, the trade can be avoided. This dynamic filtering reduces the likelihood of engaging in loss-making trades.

The improved strategy dynamically adjusts based on market conditions, effectively timing the carry trade to maximize profitability and reduce risk. By avoiding trades when the risk premium is negative, the strategy avoids scenarios where exchange rate depreciation offsets interest rate gains. Furthermore, the conditional approach can help mitigate risks during periods of heightened market volatility or safe-haven flows, where traditional carry trades often suffer losses. Over time, this results in higher average returns and improved risk-adjusted performance compared to the static carry trade.